### DISCUSSION TITLE
Persistent Room that once created can Start or Restart again & again anytime without recreating

### DISCUSSION THREAD

**Discussion by ajayalag1974** [mynaparrot/plugNmeet-server#294] [Q&A] [answered]
Hi,

I am looking for something like Persistent Room that once created can Start or Restart again & again anytime without recreating (zoom meetings have this), create room once and than just about user access token that gets expired that needs to be renewed.

CURRENT SETUP (hacky method / not ideal) -
1. Set empty_timeout to some infinite value while creating room so that room never goes into inactive state by itself.
2. Hide "End" menu item via css for all, so users cannot end room instead just can logout room.

QUESTIONS -
1. Does above hacky method have any consequences like memory / cpu being used all the time by infinitely active rooms or they are just persisted in database and cause no issue when being empty but in active state.
2. Any ideal solution already implemented for this or planned to be implemented in future ?

Thanks

**Comment by jibon57** [comment]
I don't think it's an ideal solution. In your own application you should create one static room. User will always use that link to join the session. When user will try to join during that time your system will send API request to plugNmeet for creating room. You can check the flow from here: https://www.plugnmeet.org/docs/tutorials/quick_php . We're using same pattern for our wordpress, moodle & joomla plugin.

**Reply by jibon57** [reply]
Also once user join & left the session during that time there won't be any effect regarding what value you had set for `empty_timeout`. Room will be auto ended.

**Comment by ajayalag1974** [comment]
Hi,

Your RECOMMENDED FLOW or similar flow is this -
1. Check if Room Active ?
2. If Room not Active then Create Room.
3. Generate new User Access Token.
4. Generate join url with current User Access Token and redirect.

MY FLOW is this (to avoid some extra http requests) -
1. Check if cached User Access Token not expired by decoding JWT Access token ?
2. If User Access Token is expired or not cached yet, generate new User Access Token.
3. If generating User Access Token fails with message "room is not active. create room first" (this happens only 1st time), create room with infinite empty_timeout & generate new User Access Token.
4. Generate join url with current User Access Token and redirect.

See my flow avoided 1st & 2nd step of your flow for once occuring 3rd step of my flow, generating User Access Token everytime is also avoided by caching, decoding & checking if it has not expired to reuse. All this to avoid some extra http requests.

QUESTIONS -
1. Does above hacky method have any consequences like memory / cpu being used all the time by infinitely active rooms or they are just persisted in database and cause no issue when being empty but in active state.

**Comment by ajayalag1974** [comment, marked as answer]
I used alternative flow to solve the issue - 

> HOST -
1. Create Room irrespective if its already created & active, will save request of verifying if Room is active & than create.
2. Check if cached User Access Token not expired by decoding JWT Access token ?
3. If User Access Token is expired or not cached yet, generate new User Access Token.
4. Generate join url with current User Access Token and redirect.

> VIEWERS -
1. Check if cached User Access Token not expired by decoding JWT Access token ?
2. If User Access Token is expired or not cached yet, generate new User Access Token.
3. If generating User Access Token fails with message "room is not active. create room first", don't show join link as Hosts are not active otherwise Room should have exist.
4. If User Access Token is available, generate join url with current User Access Token and redirect, might show them Room does not exist temporarily  (if token was served from cache & room is not active) until Hosts become active & room gets created.

This flow might not makes sense for somebody but for my already heavy website, this could save some resources for server / time for user.
